reactive/kotlinx-coroutines-rx2/test/SchedulerTest.kt

```diff
@@ -31,4 +38,252 @@ class SchedulerTest : TestBase() {
+        suspendCancellableCoroutine<Unit> {
+            scheduler.scheduleDirect({
+                val t1 = Thread.currentThread()
+                assertSame(t1, mainThread)
+                expect(2)
+                it.resume(Unit)
+            }, 0, TimeUnit.MILLISECONDS)
+        }
+
+        scheduler.shutdown()
+        finish(3)
+    }
+
+    @Test
+    fun testAsSchedulerWithNegativeDelay(): Unit = runTest {
+        expect(1)
+        val mainThread = Thread.currentThread()
+        val scheduler = (currentDispatcher() as CoroutineDispatcher).asScheduler()
+        suspendCancellableCoroutine<Unit> {
+            scheduler.scheduleDirect({
+                val t1 = Thread.currentThread()
+                assertSame(t1, mainThread)
+                expect(2)
+                it.resume(Unit)
+            }, -1, TimeUnit.MILLISECONDS)
+        }
+        yield()
+        finish(3)
+    }
+
+    @Test
+    fun testDisposeDuringDelay(): Unit = runTest {
+        expect(1)
+        val scheduler = (currentDispatcher() as CoroutineDispatcher).asScheduler()
+        val delayMillis = 300L
+        val disposable = scheduler.scheduleDirect({
+            expectUnreached()
+        }, delayMillis, TimeUnit.MILLISECONDS)
+        delay(100)
+        expect(2)
+        disposable.dispose()
+        delay(300)
+        yield()
+        finish(3)
+    }
+
+    @Test
+    fun testImmediateDispose(): Unit = runTest {
+        expect(1)
+        val scheduler = (currentDispatcher() as CoroutineDispatcher).asScheduler()
+        val disposable = scheduler.scheduleDirect {
+            expectUnreached()
+        }
+        disposable.dispose()
+        yield()
+        finish(2)
+    }
+
+    @Test
+    fun testAsSchedulerWorksWithSchedulerCoroutineDispatcher(): Unit = runTest {
+        expect(1)
+
+        val dispatcher = Schedulers.io().asCoroutineDispatcher()
+        val scheduler = dispatcher.asScheduler()
+        suspendCancellableCoroutine<Unit> {
+            scheduler.scheduleDirect {
+                expect(2)
+                it.resume(Unit)
+            }
+        }
+        finish(3)
+    }
+
+    @Test
+    fun tesConvertDispatcherToOriginalScheduler(): Unit = runTest {
+        expect(1)
+
+        val originalScheduler = Schedulers.io()
+        val dispatcher = originalScheduler.asCoroutineDispatcher()
+        val scheduler = dispatcher.asScheduler()
+        assertEquals(originalScheduler, scheduler)
+
+        finish(2)
+    }
+
+    @Test
+    fun tesConvertSchedulerToOriginalDispatcher(): Unit = runTest {
+        expect(1)
+
+        val originalDispatcher = currentDispatcher() as CoroutineDispatcher
+        val scheduler = originalDispatcher.asScheduler()
+        val dispatcher = scheduler.asCoroutineDispatcher()
+        assertEquals(originalDispatcher, dispatcher)
+
+        finish(2)
+    }
+
+    @Test
+    fun testExpectRxPluginsCall(): Unit = runTest {
```

Comment:
With the change I mentioned above, the ordering this test should verify is:

1. scheduleDirect is called.
2. Plugin is invoked: `expect(2)`
3. Task has not been started: `expect(3)`
3. Start the task (`DelayController.runCurrent`)
4. Plugin's wrapper Runnable is executed: `expect(4)`
5. ScheduleDirect Runnable is executed: `expect(5)`
6. `finish(6)`

Reply:
@zach-klippenstein indeed. updated.